### PR TITLE
Grouping worker

### PR DESCRIPTION
**Pull request by khaydarov**

### CODE REVIEW

**Comment by kabachook** [comment]
This pull request **introduces 2 alerts** when merging 88403796d24a9caf0643e42275c866bed66632d0 into f65c128fb3d0c7c2b21804a6a37846395a7ebece - [view on LGTM.com](https://lgtm.com/projects/g/codex-team/hawk.workers/rev/pr-b8527282b4f229789f845d8823a6cab99f150c12)

**new alerts:**

* 1 for Unused variable, import, function or class
* 1 for Hard\-coded credentials

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by kabachook** [comment]
This pull request **introduces 3 alerts** when merging c301e8b0f790f389b9357e6809224bc02b5dbe90 into f65c128fb3d0c7c2b21804a6a37846395a7ebece - [view on LGTM.com](https://lgtm.com/projects/g/codex-team/hawk.workers/rev/pr-757b0c7a835e581c64ef3a2d34da7ad89f3695ee)

**new alerts:**

* 1 for Missing &#39;this&#39; qualifier
* 1 for Unused variable, import, function or class
* 1 for Hard\-coded credentials

---

**Warning** - Automated code review for codex-team/hawk.workers will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: lib/utils.js [hunk 1]
```diff
@@ -4,3 +4,131 @@
...
+    // eslint-disable-next-line no-prototype-builtins
+    if (!objectB.hasOwnProperty(field)) {
+      continue;
+    }
+
+    const objectAItem = objectA[field];
+    const objectBItem = objectB[field];
+
+    if (!objectAItem) {
+      diffObject[field] = objectBItem;
```
Comment (kabachook) [inline lib/utils.js]: Prototype pollution

File: lib/utils.js [hunk 2]
```diff
@@ -4,3 +4,131 @@
...
+      }
+
+      if (isObject(source[key])) {
+        if (!target[key]) {
+          Object.assign(target, { [key]: {} });
+        }
+
+        deepMerge(target[key], source[key]);
+      } else {
+        Object.assign(target, { [key]: source[key] });
```
Comment (kabachook) [inline lib/utils.js]: Prototype pollution

File: lib/worker.js [hunk 3]
```diff
@@ -60,6 +60,10 @@ class Worker {
     this.tasksMap = new Map();
   }
 
+  static get delegationType() {
+    return 'grouper';
```
Comment (nikmel2803) [inline lib/worker.js]:
```suggestion
    return false;
```

File: workers/grouper/index.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  /**
+   * Start consuming messages
+   */
+  async start() {
+    await db.connect();
+    await super.start();
+
+    testDiffs();
```
Comment (nikmel2803) [inline workers/grouper/index.js]: лучше перенести в юнит-тесты

File: workers/javascript/index.js [hunk 5]
```diff
@@ -86,12 +84,11 @@ class JavascriptWorker extends tokenVerifierMixin(Worker) {
       context: event.context
     };
 
-    const insertedId = await db.saveEvent(event.projectId, {
+    await this.delegate({
```
Comment (nikmel2803) [inline workers/javascript/index.js]: производить делегирование автоматически, если у воркера определён `delegationType`

File: lib/utils.js [hunk 6]
```diff
@@ -4,3 +4,131 @@
...
+  if (!sources.length) {
+    return target;
+  }
+
+  const source = sources.shift();
+
+  if (isObject(target) && isObject(source)) {
+    for (const key in source) {
+      // eslint-disable-next-line no-prototype-builtins
+      if (!source.hasOwnProperty(key)) {
```
Comment (nikmel2803) [inline lib/utils.js]: тут вроде нет причин игнорить рекомендации eslint

Comment (nikmel2803) [reply lib/utils.js]: и выше тоже

Comment (khaydarov) [reply lib/utils.js]: image

Comment (khaydarov) [reply lib/utils.js]: webstorm почему-то ругается

File: workers/grouper/runner.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    console.log('exceptionHandler');
+    worker.logger.error(err);
+    exitHandler();
+  };
+
+  const worker = new GrouperWorker();
+
+  try {
+    await worker.start();
+    console.log(`Worker javascript started PID: ${process.pid}`);
```
Comment (kabachook) [inline workers/grouper/runner.js]:
```suggestion
    console.log(`Worker grouper started PID: ${process.pid}`);
```

File: lib/db/controller.js [hunk 8]
```diff
@@ -73,6 +73,75 @@ class Controller {
...
+  /**
+   * Inserts unique event repetition to the database
+   *
+   * @param {string|ObjectID} projectId
+   * @param {object} eventDiff
+   * @return {Promise<void>}
+   */
+  async saveRepetition(projectId, eventDiff) {
+    if (!projectId || !mongodb.ObjectID.isValid(projectId)) {
+      throw new ValidationError('Controller.saveEvent: Project ID is invalid or missed');
```
Comment (kabachook) [inline lib/db/controller.js]:
```suggestion
      throw new ValidationError('Controller.saveEvent: Project ID is invalid or missing');
```

**Pull request merged**